src/System.Private.CoreLib/shared/System/Threading/Mutex.cs

```diff
@@ -0,0 +1,218 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.IO;
+using Microsoft.Win32;
+using Microsoft.Win32.SafeHandles;
+using System.Runtime.InteropServices;
+using System.Diagnostics;
+
+namespace System.Threading
+{
+    /// <summary>
+    /// Synchronization primitive that can also be used for interprocess synchronization
+    /// </summary>
+    public sealed partial class Mutex : WaitHandle
+    {
+        private const uint AccessRights =
+            (uint)Win32Native.MAXIMUM_ALLOWED | Win32Native.SYNCHRONIZE | Win32Native.MUTEX_MODIFY_STATE;
+
+#if CORECLR && PLATFORM_UNIX
+        // Maximum file name length on tmpfs file system.
+        private const int WaitHandleNameMax = 255;
+#endif
+
+        public Mutex(bool initiallyOwned, string name, out bool createdNew)
+        {
+            VerifyNameForCreate(name);
+            CreateMutexCore(initiallyOwned, name, out createdNew);
+        }
+
+        public Mutex(bool initiallyOwned, string name)
+        {
+            VerifyNameForCreate(name);
+            CreateMutexCore(initiallyOwned, name, out _);
+        }
+
+        public Mutex(bool initiallyOwned)
+        {
+            CreateMutexCore(initiallyOwned, null, out _);
+        }
+
+        public Mutex()
+        {
+            CreateMutexCore(false, null, out _);
+        }
+
+        private Mutex(SafeWaitHandle handle)
+        {
+            SafeWaitHandle = handle;
+        }
+
+        public static Mutex OpenExisting(string name)
+        {
+            switch (OpenExistingWorker(name, out Mutex result))
+            {
+                case OpenExistingResult.NameNotFound:
+                    throw new WaitHandleCannotBeOpenedException();
+
+                case OpenExistingResult.NameInvalid:
+                    throw new WaitHandleCannotBeOpenedException(SR.Format(SR.Threading_WaitHandleCannotBeOpenedException_InvalidHandle, name));
+#if CORECLR
```

Comment:
This is unreachable code because of the previous throw. This doe not look right.